### ISSUE TITLE
[Module Proposal]: `avm-res-app-containerapp` (tf) 

### ISSUE THREAD

**Issue by kewalaka** [Azure/Azure-Verified-Modules#266]
### Check for previous/existing GitHub issues/module proposals

- [X] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [X] I have checked for that this module doesn't already exist in the module indexes; or I'm proposing the module to be migrated from CARML/TFVM.

### Bicep or Terraform?

Terraform

### Module Classification?

Resource Module

### Module Name

avm-res-app-containerapp

### Module Details

This is a module for a [container apps resource](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/container_app).

This would be the Terraform version of the Bicep equivalent here: #128 

kewalaka will work with jaredfholgate as an external contributor on this issue.

### Do you want to be the owner of this module?

No

### Module Owner's GitHub Username (handle)

jaredfholgate

### (Optional) Secondary Module Owner's GitHub Username (handle)

_No Response_

**Comment by kewalaka** [comment]
I have a starter for this, based on AVM guidance, at this location:

https://github.com/kewalaka/terraform-azurerm-avm-res-app-containerapp

I'm using AZAPI because the purpose I've made this for requires workload profile support, which is [Generally Available](https://techcommunity.microsoft.com/t5/apps-on-azure-blog/generally-available-azure-container-apps-workload-profiles-more/ba-p/3913345) but [not yet supported](https://github.com/hashicorp/terraform-provider-azurerm/issues/21747) by the terraform azurerm provider.

**Comment by mbilalamjad** [comment]
Thank you @kewalaka for the module proposal, we will find a module owner and update accordingly

**Comment by jaredfholgate** [comment]
Hi. I could take ownership of this module if nobody else on the teams is able to. Thanks

**Comment by mbilalamjad** [comment]
> Hi. I could take ownership of this module if nobody else on the team is able to. Thanks

@PmeshramPM could you please advise if it's ok for Jared to take this or an owner from Apps solution area would be needed?

**Comment by PmeshramPM** [comment]
@mbilalamjad - Yes, @jaredfholgate  can take the ownership on this module. We can always have two owners.

**Comment by mbilalamjad** [comment]
Hi @jaredfholgate,

Thanks for requesting to be an AVM module owner!

We just want to confirm **you agree to the below pages** that define what module ownership means:

- [Team Definitions & RACI](https://azure.github.io/Azure-Verified-Modules/specs/shared/team-definitions)
- [Shared Specification (Bicep & Terraform)](https://azure.github.io/Azure-Verified-Modules/specs/shared)
- [Module Support](https://azure.github.io/Azure-Verified-Modules/help-support/module-support)

Any questions or clarifications needed, let us know!

If you agree, please just **reply to this issue with the exact sentence below** (as this helps with our automation 👍):

"I CONFIRM I WISH TO OWN THIS AVM MODULE AND UNDERSTAND THE REQUIREMENTS AND DEFINITION OF A MODULE OWNER"

Thanks,

The AVM Core Team

#RR

<br>

> **NOTE**:
>
> - This Module Proposal issue **MUST remain open** until the module is fully developed, tested and published to the relevant registry. **Do NOT close** the issue before the successful publication is confirmed!
> - Once the module is fully developed, tested and published to the relevant registry, and the Module Proposal issue was closed, it **MUST remain closed**.

**Comment by jaredfholgate** [comment]
I CONFIRM I WISH TO OWN THIS AVM MODULE AND UNDERSTAND THE REQUIREMENTS AND DEFINITION OF A MODULE OWNER

**Comment by mbilalamjad** [comment]
Hi @jaredfholgate,

Thanks for confirming that you wish to own this AVM module and understand the related requirements and responsibilities!

We just want to ask you to double check a few important things and take the next steps before you start the development.

**Please use the following values explicitly as provided in the [module index](https://azure.github.io/Azure-Verified-Modules/indexes/) page**:

- For your module:
  - `ModuleName` - to name your module
  - `TelemetryIdPrefix` - to be used in your module's [telemetry](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-sfr3---category-telemetry---deploymentusage-telemetry)
- For your module's repository:
  - Repository name and folder path are defined in `RepoURL`
  - Create the GitHub teams defined in the `ModuleOwnersGHTeam` and `ModuleContributorsGHTeam` columns and grant them permissions as described [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr20---category-contributionsupport---github-teams-only).
    - In case of a Bicep module, **please let the AVM core team know** when you created these teams, and we'll help you with assigning the necessary permissions as described [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#grant-permissions---bicep).
    - In case of a Terraform module, follow the instructions [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#grant-permissions---terraform) and assign the necessary permissions.

Grant the right level of permissions for  the AVM core team and PG teams on your GitHub repo as described [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr9---category-contributionsupport---avm--pg-teams-github-repo-permissions).

You can now start the development of this module! ✅ Happy coding! 🎉

**Please respond to this comment and request a review from the AVM core team once your module is ready to be published!** Please include a link pointing to your PR if available. 🙏

Any further questions or clarifications needed, let us know!

Thanks,

The AVM Core Team

<br>

> **NOTE**:
>
> - This Module Proposal issue **MUST remain open** until the module is fully developed, tested and published to the relevant registry. **Do NOT close** the issue before the successful publication is confirmed!
> - Once the module is fully developed, tested and published to the relevant registry, and the Module Proposal issue was closed, it **MUST remain closed**.

**Comment by mbilalamjad** [comment]
> I have a starter for this, based on AVM guidance, at this location:
> 
> https://github.com/kewalaka/terraform-azurerm-avm-res-app-containerapp
> 
> I'm using AZAPI because the purpose I've made this for requires workload profile support, which is [Generally Available](https://techcommunity.microsoft.com/t5/apps-on-azure-blog/generally-available-azure-container-apps-workload-profiles-more/ba-p/3913345) but [not yet supported](https://github.com/hashicorp/terraform-provider-azurerm/issues/21747) by the terraform azurerm provider.

@jaredfholgate please work with @kewalaka, you would have to own this module as Microsoft FTE and can have @kewalaka as co-owner and leverage the great work he's already done for this proposed module
CC @PmeshramPM

**Comment by jaredfholgate** [comment]
> Hi @jaredfholgate,
> 
> Thanks for confirming that you wish to own this AVM module and understand the related requirements and responsibilities!
> 
> We just want to ask you to double check a few important things and take the next steps before you start the development.
> 
> **Please use the following values explicitly as provided in the [module index](https://azure.github.io/Azure-Verified-Modules/indexes/) page**:
> 
> * For your module:
>   
>   * `ModuleName` - to name your module
>   * `TelemetryIdPrefix` - to be used in your module's [telemetry](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-sfr3---category-telemetry---deploymentusage-telemetry)
> * For your module's repository:
>   
>   * Repository name and folder path are defined in `RepoURL`
>   * Create the GitHub teams defined in the `ModuleOwnersGHTeam` and `ModuleContributorsGHTeam` columns and grant them permissions as described [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr20---category-contributionsupport---github-teams-only).
>     
>     * In case of a Bicep module, **please let the AVM core team know** when you created these teams, and we'll help you with assigning the necessary permissions as described [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#grant-permissions---bicep).
>     * In case of a Terraform module, follow the instructions [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#grant-permissions---terraform) and assign the necessary permissions.
> 
> Grant the right level of permissions for the AVM core team and PG teams on your GitHub repo as described [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr9---category-contributionsupport---avm--pg-teams-github-repo-permissions).
> 
> You can now start the development of this module! ✅ Happy coding! 🎉
> 
> **Please respond to this comment and request a review from the AVM core team once your module is ready to be published!** Please include a link pointing to your PR if available. 🙏
> 
> Any further questions or clarifications needed, let us know!
> 
> Thanks,
> 
> The AVM Core Team
> 
> 
> > **NOTE**:
> > 
> > * This Module Proposal issue **MUST remain open** until the module is fully developed, tested and published to the relevant registry. **Do NOT close** the issue before the successful publication is confirmed!
> > * Once the module is fully developed, tested and published to the relevant registry, and the Module Proposal issue was closed, it **MUST remain closed**.

Ready for review.

PRs:

- https://github.com/Azure/terraform-azurerm-avm-res-app-containerapp/pull/2
- https://github.com/Azure/terraform-azurerm-avm-res-app-containerapp/pull/1

**Comment by jaredfholgate** [comment]
@mbilalamjad I may have misunderstood the message. If you don't want me to request the changes until the code is ready, let me know and I will cancel the PRs.

**Comment by jaredfholgate** [comment]
> @jaredfholgate Jared Holgate FTE, @kewalaka, this issue has not had any activity in the last **3 weeks**. Please feel free to reach out to the AVM core team should you have any questions or need any help with the development of this module.

I was in touch with Kewalaka yesterday to discuss timings and my availability. Will hopefully migrate code next week and then work on it properly in January.

**Comment by davidkarlsen** [comment]
can https://github.com/Azure/terraform-azure-container-apps/ be used as a starting point?

**Comment by kewalaka** [comment]
> can https://github.com/Azure/terraform-azure-container-apps/ be used as a starting point?

hi @davidkarlsen - [my version](https://github.com/kewalaka/terraform-azurerm-avm-res-app-containerapp), mentioned at the start of the thread, is based on this.  There's a couple implementations in there, both the AzureRM & AzAPI versions.  Again similar to app managed environments, due to Container Apps being relatively new & having frequent changes, I've used AzAPI.  I actually needed workload profiles, which at the time wasn't supported on AzureRM even tho it was GA (it was added in the last few weeks, but [still has bugs](https://github.com/hashicorp/terraform-provider-azurerm/issues/24198)).

In any case, how the module achieves its outcome shouldn't matter from the prospective of a module consumer (to TLDR guidance from AVM - they're complex on the inside to allow for flexibility on the outside) - so my opinion is to stay with AzAPI.

**Comment by kewalaka** [comment]
& as an aside, there is a conversion utility - [azapi2azurerm](https://github.com/Azure/azapi2azurerm) shoudl that position change.  Microsoft have made it clear in recent community calls that AzAPI is a fully supported provider not simply a bridge technology.

My experience - both azapi & azurerm have bugs in the container apps space, but the azapi ones I've run into more relate to "noise" in the plan vs functional issues.  In general, I've had a good experience to date with the azapi version, I'm using it quite a bit in a project.

**Comment by mbilalamjad** [comment]
Thanks @kewalaka and @davidkarlsen for your comments, will let @jaredfholgate advise..

**Comment by kewalaka** [comment]
@jaredfholgate I've done some work recently to re-align my original module so that the variable block follows the same pattern and naming as the azurerm interface, with the intent that the consumption of the module should be the same, just the implementation within that differs.

See latest changes here - https://github.com/kewalaka/terraform-azapi-res-app-containerapp/

I've since thought that the AVM resource should focus on the azurerm provider, so for now I've renamed my project to 'azapi'.  Not sure how to best approach this.

Below are some thoughts (& other suggestions welcome) ;

- should the azurerm version track the capabilities of the official AzureRM provider?  Perhaps azapi can be used sparingly to work around azurerm issues, but not for the parent resource?

..or..

- Can the azapi version be used for the parent resource so long as the variables block follows the same pattern/interface as the azurerm equivalent resource - as per my link above.


Also:

- Should we have a feature flag in the module to switch between azurerm & azapi as the consumer desires *within the same module*?
...or ..
- Should it be separate resource module - i.e. starting terraform-azapi & terraform-azurerm.  Whilst that seems reasonable, this seems like a misuse of the naming convention 😊


I understand that in general we want to use the azurerm provider - it's just with some of these resources there have been significant gaps in recent memory that make this undesirable for several use cases.

cc @matt-FFFFFF as discussed.

**Comment by jaredfholgate** [comment]
Hi All. Since taking ownership of this module it has come to light that there is a pre-existing Microsoft maintained ACA module. 

https://github.com/Azure/terraform-azure-container-apps

@lonegunmanb and @jiaweitao001 are the owners of this module. I have asked them if they would prefer to the be the owners and continue to maintain this module moving forwards, given their ACA expertise. They do, so I will transfer ownership. They will then migrate their existing code to the new repo and bring it in line with AVM.

@mbilalamjad I can update the owners groups, but I guess you my need to update the ownership data elsewhere.

@kewalaka I appreciate that your version of the module includes extra working GA features that are not available in azurerm yet. My suggestion to the new owners is that they accept a PR to add the extra functionality using the AzAPI resources as needed until azurerm equivalent is available. So a hybrid approach.

I hope this is ok with everyone and apologies for the amount of time it took to get to this point.

**Comment by kewalaka** [comment]
@jaredfholgate makes sense I was coming to the same conclusion too, which is why i've decided to rename mine to make it clearer it is azapi-based.  I am tracking the module inputs specification of the azurerm version so that people can switch between the two easily.  I find AzAPI more fiddly to diagnose so I would prefer switching to an AzureRM version long term.

On the [mentioned ](https://github.com/Azure/terraform-azure-container-apps) repo, these are a few observations:

- The container apps environment already exists as a [separate resource](https://github.com/azure/terraform-azurerm-avm-res-app-managedenvironment) 
- There will be a few breaking changes as it comes across - i note the existing module supports making N container apps, rather than 1 - which is not to spec - in part as it makes some of the internal logic more fiddly.
- There's the expected minor naming & interface changes needed to align with the spec & shared interfaces in AVM, which will introduce breaking changes for people migrating between the two.

@lonegunmanb & @jiaweitao001 - happy to assist with this if there is appetite to do this work in an open forum?

**Comment by kewalaka** [comment]
I see this hasn't yet been transferred to @lonegunmanb and @jiaweitao001 , can this be actioned please?

I'm happy to support the work, however the official approach has blocking issues for me (relating to fixes required upstream in the AzureRM provider), so therefore for the time being I'm continuing this approach:

https://github.com/kewalaka/terraform-azapi-res-app-containerapp

This seeks to provide inputs & outputs that mirror azurerm, whilst using AzAPI internally.

I note this isn't (quite) feature complete yet, and I realise it probably doesn't fit the desired approach, but sharing in case.

**Comment by mbilalamjad** [comment]
Requested @jaredfholgate to share an update on this

**Comment by v-xuto** [comment]
Related to issue: https://github.com/Azure/Azure-Verified-Modules/issues/261
CC: @jongio

**Comment by v-xuto** [comment]
@jaredfholgate - Do you have an ETA for this?

**Comment by jaredfholgate** [comment]
> @jaredfholgate - Do you have an ETA for this?

Hi @lonegunmanb is the owner of this module. I appreciate my name is still against it, this is due to an admin issue.

**Comment by mbilalamjad** [comment]
Requesting @prjelesi to help transfer ownership to @lonegunmanb once he confirms.

**Comment by lonegunmanb** [comment]
I CONFIRM I WISH TO OWN THIS AVM MODULE AND UNDERSTAND THE REQUIREMENTS AND DEFINITION OF A MODULE OWNER

**Comment by frandariem** [comment]
Hi team, could you please provide an update? We are very interested in using this module.

**Comment by lonegunmanb** [comment]
I'd like to migrate [terraform-azure-container-apps](https://github.com/Azure/terraform-azure-container-apps) to AVM, I hope I can finish this task in next week.

**Comment by v-xuto** [comment]
@lonegunmanb   When will this module be published?

**Comment by lonegunmanb** [comment]
The initial review of this module is complete and the module has been published to the registry. Requesting AVM team to close this module proposal and mark the module available in the module index.

**Comment by lonegunmanb** [comment]
@v-xuto @kewalaka This module is now available at [here](https://registry.terraform.io/modules/Azure/avm-res-app-containerapp/azurerm/latest)

**Comment by v-xuto** [comment]
@lonegunmanb Thank you for your contribution. We will start testing.
